### DISCUSSION TITLE
Can you have a standard extension template? Our team wants to extend MYSQL module by ourselves

### DISCUSSION THREAD

**Discussion by JackSlowFucck** [pgsty/pigsty#228] [General]
Can you have a standard extension template? Our team wants to extend MYSQL module by ourselves

**Comment by Vonng** [comment]
Pigsty's modules are standard ansible roles & playbooks. However, it does have some additional conventions above it.

Several modules can be used as examples:

PGSQL

REDIS

MINIO

ETCD



Take `PGSQL` as an example:

The API is defined in [`[roles/pgsql/default/main.yml](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/defaults/main.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/defaults/main.yml)

The Implementation lies in [`[roles/pgsql/tasks/main.yml](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/main.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/main.yml)



To download extra packages from upstream,  you may have to add your upstream repo & related rpm names in [`[roles/repo/tasks/default/main.yml](https://github.com/Vonng/pigsty/blob/master/roles/repo/defaults/main.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/repo/defaults/main.yml).

To register a monitoring target to Prometheus, check [`[roles/pgsql/tasks/register_prometheus.yml](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_prometheus.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_prometheus.yml) for detail.

 To register a data source to Grafana, check [`[roles/pgsql/tasks/register_grafana.yml](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_grafana.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_grafana.yml) for detail.

 To add a data source to Grafana, check [`[roles/pgsql/tasks/register_grafana.yml](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_grafana.yml)`](https://github.com/Vonng/pigsty/blob/master/roles/pgsql/tasks/register_grafana.yml) for detail.

To collect new logs, the promtail config template should be adjusted: [`[roles/node_monitor/templates/promtail.yml.j2](https://github.com/Vonng/pigsty/blob/master/roles/node_monitor/templates/promtail.yml.j2)`](https://github.com/Vonng/pigsty/blob/master/roles/node_monitor/templates/promtail.yml.j2)
